### PR TITLE
Operator Bugfix: Clean out empty volumeClaimTemplate struct if set

### PR DESCRIPTION
**Pull request by bennerv**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10229043

### What this PR does / why we need it:
If the `openshift-monitoring` `cluster-monitoring-config` configmap has a volumeClaimTemplate in prometheusK8s or alertmanagerMain set to: `volumeClaimTemplate: {}` the ARO operator won't clean it up and the monitoring clusteroperator will be bad.  

### Test plan for issue:

Updated e2e and unit tests for it and ran it against the shared cluster.  

### Is there any documentation that needs to be updated for this PR?

No, but __NOTE__.  The aro operator will update the configmap, and it will take some time (I've seen 5 min+) for the monitoring operator to update the prometheus/alertmanager resources.

### CODE REVIEW

**Comment by mjudeikis** [comment]
@bennerv what about taking this step forward? https://github.com/mjudeikis/ARO-RP/commit/d2ebf9595f7bd16113098ebf5193e46cbb101ed7 ?

File: pkg/operator/controllers/monitoring/monitoring_controller.go [hunk 1]
```diff
@@ -89,33 +117,23 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
...
+	configData.PrometheusK8s.Retention = nil
+	configData.PrometheusK8s.VolumeClaimTemplate = nil
 
 	var b []byte
 	err = codec.NewEncoderBytes(&b, r.jsonHandle).Encode(configData)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
+	changed := !bytes.Equal(b, configDataJSON)
```
Comment (m1kola) [inline pkg/operator/controllers/monitoring/monitoring_controller.go:131]: Not sure that comparing json strings is reliable: you can have same data, but marshaled in different order.

If we want to exit early if `!changed` then we should be checking on go value, I think. Similarly to what we had before:

```go
if configData.PrometheusK8s.Retention != "" {
    configData.PrometheusK8s.Retention = ""
    changed = true
}

if configData.PrometheusK8s.VolumeClaimTemplate != nil {
    configData.PrometheusK8s.VolumeClaimTemplate = nil
    changed = true
}

if configData.AlertManagerMain.VolumeClaimTemplate != nil {
    configData.AlertManagerMain.VolumeClaimTemplate = nil
    changed = true
}
```

Is there a better option?

File: pkg/operator/controllers/monitoring/monitoring_controller_test.go [hunk 2]
```diff
@@ -111,11 +100,73 @@ alertmanagerMain:
...
+			},
+			wantConfig: `
+alertmanagerMain:
+  extraField: alertmanager
+prometheusK8s:
+  bugs: not-here
+`,
+		},
+		{
+			name: "empty volumeClaimTemplate cleared out part 2",
```
Comment (m1kola) [inline pkg/operator/controllers/monitoring/monitoring_controller_test.go:134]: Unless I miss something, `"empty volumeClaimTemplate cleared out part 2"` case covers everything we need. We can probably drop `"empty volumeClaimTemplate struct is cleared out"` case. What do you think?

```suggestion
			name: "empty volumeClaimTemplate cleared out",
```

File: pkg/operator/controllers/monitoring/monitoring_controller_test.go [hunk 3]
```diff
@@ -165,3 +216,109 @@ somethingElse:
...
 	}
 }
+
+func TestReconcilePVC(t *testing.T) {
+	log := logrus.NewEntry(logrus.StandardLogger())
+	tests := []struct {
+		name       string
+		reconciler func() *Reconciler
+		want       []corev1.PersistentVolumeClaim
+		wantErr    error
```
Comment (m1kola) [inline pkg/operator/controllers/monitoring/monitoring_controller_test.go:226]: We do not seem to use `wantErr`.

File: pkg/operator/controllers/monitoring/monitoring_controller_test.go [hunk 4]
```diff
@@ -165,3 +216,109 @@ somethingElse:
 		})
 	}
 }
+
+func TestReconcilePVC(t *testing.T) {
+	log := logrus.NewEntry(logrus.StandardLogger())
+	tests := []struct {
+		name       string
+		reconciler func() *Reconciler
```
Comment (m1kola) [inline pkg/operator/controllers/monitoring/monitoring_controller_test.go:224]: Should we replace this (and refactor the older test for consistency) with `*fake.Clientset`? Because the only thing that chances in reconciler is fake client.
We can instantiate the reconciler itself in the test body.

**Review by m1kola — DISMISSED** [review]
Looks good overall. Just few minor things in the controller and few suggestions for the tests.

File: pkg/operator/controllers/monitoring/monitoring_controller_test.go (hunk 2 above)
Comment (bennerv) [reply pkg/operator/controllers/monitoring/monitoring_controller_test.go:134]: forgot to remove that one - thought I hit an edge case so I threw in another additional test.  
Good catch will update.

File: pkg/operator/controllers/monitoring/monitoring_controller_test.go (hunk 3 above)
Comment (bennerv) [reply pkg/operator/controllers/monitoring/monitoring_controller_test.go]: Removed

File: pkg/operator/controllers/monitoring/monitoring_controller.go [hunk 5]
```diff
@@ -38,21 +41,17 @@ type Config struct {
 	api.MissingFields
 	PrometheusK8s struct {
 		api.MissingFields
-		Retention           string `json:"retention,omitempty"`
-		VolumeClaimTemplate struct {
-			api.MissingFields
+		Retention           *string `json:"retention,omitempty"`
+		VolumeClaimTemplate *struct {
 		} `json:"volumeClaimTemplate,omitempty"`
```
[the remark this answers is not in the document]
Comment (bennerv) [reply pkg/operator/controllers/monitoring/monitoring_controller.go]: Updated.  It's using a pointer to `json.RawMessage` as there has to be a distinction between an empty RawMessage and a not set `RawMessage`.

File: pkg/operator/controllers/monitoring/monitoring_controller.go (hunk 1 above)
Comment (bennerv) [reply pkg/operator/controllers/monitoring/monitoring_controller.go]: Yes it pretty much always reported as changed == true.  Now that we're using pointers for `VolumeClaimTemplate` fields and the `json.RawMessage` we can go back to this.  

Updated.

File: pkg/operator/controllers/monitoring/monitoring_controller_test.go (hunk 4 above)
Comment (bennerv) [reply pkg/operator/controllers/monitoring/monitoring_controller_test.go]: Refactored.  Should be below instead of reconcile function.  
```
kubernetescli kubernetes.Interface
```

**Review by mjudeikis — APPROVED** [review]
